File: src/cacheRiskCommands.ts
import * as fs from 'fs'
import * as path from 'path'
import { claudeProjectsDirs } from './logReader'
import { classifySlashCommand, type CacheRiskKind, type MutationCertainty } from './shared/cacheRiskKinds'

// Re-exported (never re-declared) so host callers have one import site and check-mirrors keeps the
// dashboard honest — the vocabulary itself lives in src/shared/cacheRiskKinds.ts.
export { classifySlashCommand, CACHE_RISK_STYLE } from './shared/cacheRiskKinds'
export type { CacheRiskKind, MutationCertainty } from './shared/cacheRiskKinds'

// Cache-risk slash commands, read straight out of the Claude Code transcript (TRDD-EYA3X5MQ).
//
// WHY a transcript scan and not a hook: there is no plugin-reload hook event, built-in slash
// commands do not fire UserPromptSubmit, and `ConfigChange` was measured and REFUTED as a reload
// signal (an explicit /reload-plugins emitted ZERO hook events; 40,858 stored records hold
// ConfigChange 0 times). But Claude Code DOES persist every built-in command it runs as a normal
// `type: "user"` transcript entry whose content is a `<command-name>…</command-name>` block — so
// the exact command, its args, its timestamp and its session are already on disk, retroactively,
// for the whole history. That is ground truth; the catalog co-churn heuristic is a fallback for
// turns with no transcript (it over-counted this project 102 vs 69 actual).
//
// WHY `mutation` exists: invoking a command is not the same as changing anything. `/reload-plugins`
// and `/reload-skills` are actions — they always re-register catalogs. Bare `/plugin` and `/mcp`
// open an interactive menu the user may simply close, and `/model` may be re-selected to the same
// model. Those are 'ambiguous': the caller must corroborate them against real cost/catalog churn
// before charging a cache break to them, or it repeats the over-count the inference already made.

// NO catalog sizes here, deliberately. `/reload-plugins` prints
// `Reloaded: 34 plugins · 117 skills · 75 agents · 27 hooks · …` into the live conversation, which
// makes it tempting to parse for "how big was this reload". It is NOT on disk: a machine-wide scan
// found only 14 records containing that line inside a `<local-command-stdout>` tag, and ZERO of
// them are command entries — every one is a QUOTE (an assistant message, a tool_result, a
// compaction summary). Claude Code persists the command block and drops the stdout. Do not add a
// catalogs field back without first proving a real command entry carries one.
export interface CacheRiskCommand {
  /** Epoch ms of the command entry. */
  ts: number
  /** Session id the command was typed in (transcript `sessionId`). */
  session?: string
  /** The literal command, e.g. `/reload-plugins`. */
  command: string
  /** Raw `<command-args>` text, when non-empty. */
  args?: string
  kind: CacheRiskKind
  mutation: MutationCertainty
}

const RE_NAME = /^(?:\s*<local-command-caveat>[\s\S]*?<\/local-command-caveat>)?\s*<command-name>([^<]*)<\/command-name>/
const RE_ARGS = /<command-args>([^<]*)<\/command-args>/

/** Split a transcript entry's text into its command parts. Undefined when it is not a command. */
export function parseCommandBlock(text: string): { name: string; args?: string } | undefined {
  const name = RE_NAME.exec(text)?.[1]?.trim()
  if (!name) return undefined
  const args = RE_ARGS.exec(text)?.[1]?.trim()
  return { name, args: args || undefined }
}

/** A parsed transcript line, narrowed to the fields this module needs. */
interface TranscriptEntry {
  type?: unknown
  timestamp?: unknown
  sessionId?: unknown
  message?: { content?: unknown }
}

/**
 * Turn one transcript entry into a cache-risk command, or undefined. Only `type: "user"` entries
 * whose content is a STRING carry a command block — a tool_result is an array, and treating one as
 * a command would let any transcript that merely QUOTES `<command-name>` (an assistant explaining
 * this very feature, for instance) inflate the count.
 */
export function extractCacheRiskCommand(entry: TranscriptEntry): CacheRiskCommand | undefined {
  if (entry.type !== 'user') return undefined
  const content = entry.message?.content
  if (typeof content !== 'string') return undefined
  const block = parseCommandBlock(content)
  if (!block) return undefined
  const cls = classifySlashCommand(block.name, block.args)
  if (!cls) return undefined
  const ts = typeof entry.timestamp === 'string' ? Date.parse(entry.timestamp) : NaN
  if (!Number.isFinite(ts)) return undefined
  const cmd: CacheRiskCommand = {
    ts,
    session: typeof entry.sessionId === 'string' ? entry.sessionId : undefined,
    command: block.name.trim(),
    kind: cls.kind,
    mutation: cls.mutation,
  }
  if (block.args) cmd.args = block.args
  return cmd
}

export interface ScanOptions {
  /** Only consider commands at or after this epoch ms (also skips whole files by mtime). */
  sinceMs?: number
  /** Restrict to these kinds. */
  kinds?: ReadonlyArray<CacheRiskKind>
  /** Newest-first cap on the returned list. */
  limit?: number
  /** Transcript roots; defaults to Claude Code's own project dirs. */
  dirs?: ReadonlyArray<string>
}

/** Every `*.jsonl` under the given roots, one level deep (Claude Code's layout: <root>/<slug>/*.jsonl). */
function transcriptFiles(dirs: ReadonlyArray<string>): string[] {
  const out: string[] = []
  for (const root of dirs) {
    let slugs: string[]
    try { slugs = fs.readdirSync(root) } catch { continue }
    for (const slug of slugs) {
      const dir = path.join(root, slug)
      let names: string[]
      try { names = fs.readdirSync(dir) } catch { continue }
      for (const n of names) if (n.endsWith('.jsonl')) out.push(path.join(dir, n))
    }
  }
  return out
}

/**
 * Scan transcripts for cache-risk commands, newest first.
 *
 * Two cheap filters keep a full-history scan (12k+ files) affordable: skip a file whose mtime
 * predates `sinceMs`, and skip a file whose text has no `<command-name>` at all — only then parse,
 * and only the lines that contain the marker.
 */
export function scanCacheRiskCommands(opts: ScanOptions = {}): CacheRiskCommand[] {
  const dirs = opts.dirs ?? claudeProjectsDirs()
  const kinds = opts.kinds ? new Set(opts.kinds) : undefined
  const found: CacheRiskCommand[] = []
  for (const file of transcriptFiles(dirs)) {
    if (opts.sinceMs !== undefined) {
      try { if (fs.statSync(file).mtimeMs < opts.sinceMs) continue } catch { continue }
    }
    let text: string
    try { text = fs.readFileSync(file, 'utf8') } catch { continue }
    if (!text.includes('<command-name>')) continue
    for (const line of text.split('\n')) {
      if (!line.includes('<command-name>')) continue
      let entry: TranscriptEntry
      try { entry = JSON.parse(line) as TranscriptEntry } catch { continue }
      const cmd = extractCacheRiskCommand(entry)
      if (!cmd) continue
      if (opts.sinceMs !== undefined && cmd.ts < opts.sinceMs) continue
      if (kinds && !kinds.has(cmd.kind)) continue
      found.push(cmd)
    }
  }
  found.sort((a, b) => b.ts - a.ts)
  return opts.limit !== undefined ? found.slice(0, opts.limit) : found
}

File: rust-core/crates/agentlens-core/src/cache_risk_commands.rs
//! Cache-risk slash commands read straight out of the Claude Code transcripts (TRDD-DMWOBWFH
//! row 9) — ports src/cacheRiskCommands.ts + the src/shared/cacheRiskKinds.ts classifier.
//!
//! WHY a transcript scan and not a hook: there is no plugin-reload hook event, built-ins do
//! not fire UserPromptSubmit, and ConfigChange was measured and REFUTED as a reload signal —
//! but Claude Code persists every built-in command as a `type:"user"` entry whose content is a
//! `<command-name>…</command-name>` block, retroactively for the whole history. `mutation`
//! distinguishes commands that ALWAYS mutate the prefix from menu-openers that only might;
//! /fork, /subtask, /branch and /resume are DELIBERATELY absent (their cost lands in a
//! DIFFERENT context than the session the command was typed in — classifying them here would
//! bill a child context's cold start to the parent's next turn).

use std::path::PathBuf;

use indexmap::IndexMap;
use serde_json::{json, Map, Value};

use crate::summarize::helpers::{num, parse_iso_ms};

/// classifySlashCommand — (kind, mutation) or None for commands that do not touch the typed-in
/// session's cached prefix. Pure and table-driven: the vocabulary lives in exactly one place.
pub fn classify_slash_command(name: &str, args: Option<&str>) -> Option<(&'static str, &'static str)> {
    let cmd = name.trim().to_lowercase();
    let a = args.unwrap_or("").trim().to_owned();
    match cmd.as_str() {
        "/reload-plugins" => Some(("PLUGINS_RELOADED", "certain")),
        "/reload-skills" => Some(("SKILLS_RELOADED", "certain")),
        "/login" | "/logout" => Some(("ACCOUNT_SWITCHED", "certain")),
        "/compact" => Some(("COMPACTION", "certain")),
        "/clear" => Some(("CLEAR", "certain")),
        // Menu-driven commands: an argument means the user named the mutation outright; bare
        // invocation only OPENS the picker and may change nothing at all.
        "/plugin" | "/plugins" => {
            if a.is_empty() {
                return Some(("PLUGIN_CHANGED", "ambiguous"));
            }
            // `/plugin plugin update x` appears in real transcripts — strip one redundant
            // leading `plugin` before testing the verb.
            static STRIP: std::sync::OnceLock<regex::Regex> = std::sync::OnceLock::new();
            static MUTATING: std::sync::OnceLock<regex::Regex> = std::sync::OnceLock::new();
            let strip = STRIP.get_or_init(|| regex::Regex::new(r"(?i)^plugins?\s+").expect("static regex"));
            let mutating =
                MUTATING.get_or_init(|| regex::Regex::new(r"(?i)^(install|uninstall|remove|enable|disable|update|marketplace)\b").expect("static regex"));
            if mutating.is_match(&strip.replace(&a, "")) {
                Some(("PLUGIN_CHANGED", "certain"))
            } else {
                None
            }
        }
        "/mcp" => Some(("MCP_SERVER_TOGGLE", "ambiguous")),
        "/model" => Some(("MODEL_SWITCHED", if a.is_empty() { "ambiguous" } else { "certain" })),
        "/effort" => Some(("EFFORT_CHANGED", if a.is_empty() { "ambiguous" } else { "certain" })),
        _ => None,
    }
}

/// parseCommandBlock — the command name (+ non-empty args) out of a transcript entry's text.
pub fn parse_command_block(text: &str) -> Option<(String, Option<String>)> {
    static RE_NAME: std::sync::OnceLock<regex::Regex> = std::sync::OnceLock::new();
    static RE_ARGS: std::sync::OnceLock<regex::Regex> = std::sync::OnceLock::new();
    let re_name = RE_NAME.get_or_init(|| {
        regex::Regex::new(r"(?s)^(?:\s*<local-command-caveat>.*?</local-command-caveat>)?\s*<command-name>([^<]*)</command-name>")
            .expect("static regex")
    });
    let re_args = RE_ARGS.get_or_init(|| regex::Regex::new(r"<command-args>([^<]*)</command-args>").expect("static regex"));
    let name = re_name.captures(text)?.get(1)?.as_str().trim().to_owned();
    if name.is_empty() {
        return None;
    }
    let args = re_args.captures(text).and_then(|c| c.get(1)).map(|m| m.as_str().trim().to_owned()).filter(|a| !a.is_empty());
    Some((name, args))
}

/// extractCacheRiskCommand — only `type:"user"` entries whose content is a STRING carry a
/// command block (a tool_result is an array; treating one as a command would let a transcript
/// that merely QUOTES `<command-name>` inflate the count). The wire row is
/// `{ts, session?, command, args?, kind, mutation}` — conditional keys mirrored.
pub fn extract_cache_risk_command(entry: &Value) -> Option<Value> {
    if entry.get("type").and_then(Value::as_str) != Some("user") {
        return None;
    }
    let content = entry.get("message").and_then(|m| m.get("content")).and_then(Value::as_str)?;
    let (name, args) = parse_command_block(content)?;
    let (kind, mutation) = classify_slash_command(&name, args.as_deref())?;
    // Transcript timestamps are ISO (Date.parse there, parse_iso_ms here — agree on ISO).
    let ts = entry.get("timestamp").and_then(Value::as_str).and_then(parse_iso_ms)?;
    let mut cmd = Map::new();
    cmd.insert("ts".into(), num(ts));
    if let Some(sid) = entry.get("sessionId").and_then(Value::as_str) {
        cmd.insert("session".into(), Value::from(sid));
    }
    cmd.insert("command".into(), Value::from(name.trim()));
    cmd.insert("kind".into(), Value::from(kind));
    cmd.insert("mutation".into(), Value::from(mutation));
    if let Some(a) = args {
        // `if (block.args) cmd.args = block.args` runs AFTER the literal — args serializes
        // LAST, whatever order the TS interface declares.
        cmd.insert("args".into(), Value::from(a));
    }
    Some(Value::Object(cmd))
}

/// transcriptFiles — every `*.jsonl` one level deep (<root>/<slug>/*.jsonl). Sorted for
/// determinism (the TS walks raw readdir order — observable only in same-ts tie order, which
/// no frozen shape pins).
pub fn transcript_files(dirs: &[PathBuf]) -> Vec<PathBuf> {
    let mut out = Vec::new();
    for root in dirs {
        let Ok(slugs) = std::fs::read_dir(root) else { continue };
        let mut slugs: Vec<PathBuf> = slugs.flatten().map(|e| e.path()).collect();
        slugs.sort();
        for slug in slugs {
            let Ok(names) = std::fs::read_dir(&slug) else { continue };
            let mut files: Vec<PathBuf> =
                names.flatten().map(|e| e.path()).filter(|p| p.to_string_lossy().ends_with(".jsonl")).collect();
            files.sort();
            out.extend(files);
        }
    }
    out
}

/// scanCacheRiskCommands — newest first. Two cheap filters keep a full-history scan (12k+
/// files) affordable: skip a file whose mtime predates sinceMs, and skip a file whose text has
/// no `<command-name>` at all — only then parse, and only the marker lines.
pub fn scan_cache_risk_commands(dirs: &[PathBuf], since_ms: Option<f64>, kinds: Option<&[String]>, limit: Option<usize>) -> Vec<Value> {
    let mut found: Vec<Value> = Vec::new();
    for file in transcript_files(dirs) {
        if let Some(since) = since_ms {
            let Ok(meta) = std::fs::metadata(&file) else { continue };
            let mtime = meta
                .modified()
                .ok()
                .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
                .map(|d| d.as_secs_f64() * 1000.0)
                .unwrap_or(0.0);
            if mtime < since {
                continue;
            }
        }
        let Ok(bytes) = std::fs::read(&file) else { continue };
        let text = String::from_utf8_lossy(&bytes);
        if !text.contains("<command-name>") {
            continue;
        }
        for line in text.split('\n') {
            if !line.contains("<command-name>") {
                continue;
            }
            let Ok(entry) = serde_json::from_str::<Value>(line) else { continue };
            let Some(cmd) = extract_cache_risk_command(&entry) else { continue };
            if let Some(since) = since_ms {
                if cmd.get("ts").and_then(Value::as_f64).unwrap_or(f64::NAN) < since {
                    continue;
                }
            }
            if let Some(kinds) = kinds {
                if !kinds.iter().any(|k| Some(k.as_str()) == cmd.get("kind").and_then(Value::as_str)) {
                    continue;
                }
            }
            found.push(cmd);
        }
    }
    // `b.ts - a.ts`, stable — ties keep scan order.
    found.sort_by(|a, b| {
        let t = |v: &Value| v.get("ts").and_then(Value::as_f64).unwrap_or(f64::NAN);
        t(b).partial_cmp(&t(a)).unwrap_or(std::cmp::Ordering::Equal)
    });
    if let Some(l) = limit {
        found.truncate(l);
    }
    found
}

/// The GET /api/cache-risk-commands body (server.ts:3465). Scan unlimited, then slice — the
/// response states the TRUE total (a capped list whose count silently equals the cap reads as
/// "that is all there was"); per-kind counts come from the FULL window, not the capped page.
pub fn cache_risk_commands_response(dirs: &[PathBuf], now_ms: f64, window_hours: f64, kinds: Option<&[String]>, limit: usize) -> Value {
    let all = scan_cache_risk_commands(dirs, Some(now_ms - window_hours * 3_600_000.0), kinds, None);
    let commands: Vec<Value> = all.iter().take(limit).cloned().collect();
    let mut by_kind: IndexMap<String, u64> = IndexMap::new();
    for c in &all {
        *by_kind.entry(c.get("kind").and_then(Value::as_str).unwrap_or("").to_owned()).or_insert(0) += 1;
    }
    let mut bk = Map::new();
    for (k, n) in by_kind {
        bk.insert(k, Value::from(n));
    }
    json!({
        "windowHours": num(window_hours),
        "total": all.len(),
        "count": commands.len(),
        "truncated": all.len() > commands.len(),
        "byKind": bk,
        "commands": commands,
    })
}
